File: src/MappingEdu.Web.UI/Client/App/Repositories/Entity/Entity.ts
// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

// ****************************************************************************
// Module app.repositories.entity
//

var m = angular.module('app.repositories.entity', [
    'restangular',
     'app.repositories.entity.hint']);


// ****************************************************************************
// Interface Entity
//
 
interface IEntity extends restangular.IElement {
    //TODO
}

// ****************************************************************************
// Interface IEntityRepository
//

interface IEntityRepository {
    find(mappingProjectId: string, id: string): angular.IPromise<any>;
    getFirstLevelEntities(mappedSystemId: string): angular.IPromise<any>;
    remove(id: string): angular.IPromise<any>;
    hint: IEntityHintRepository;
}

// ****************************************************************************
// Entity repository
//

m.factory('app.repositories.entity', ['Restangular', 'app.repositories.entity.hint',  (restangular: restangular.IService, hint: IEntityHintRepository) => {

    restangular.extendModel('Entity', (model: IEntity) => {
        return model;
    });

    // methods

    function find(mappingProjectId: string, id: string) {
        return restangular.one('Entity/' + mappingProjectId, id).get();
    }

    function getFirstLevelEntities(mappedSystemId) {
        return restangular.one(`Entity/${mappedSystemId}/first-level`).get();
    }

    function remove(id: string) {
        return restangular.one('Entity', id).remove();
    }

    var repository: IEntityRepository = {
        find: find,
        getFirstLevelEntities: getFirstLevelEntities,
        remove: remove,
        hint: hint
    };

    return repository;
}]); 

File: src/MappingEdu.Core.Domain/Entity.cs
// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

using System;

namespace MappingEdu.Core.Domain
{
    public abstract class Entity : IEntity
    {
        public string CreateBy { get; set; }

        public bool HasId
        {
            get { return Id != default(Guid); }
        }

        protected abstract Guid Id { get; }

        public Guid? CreateById { get; set; }

        public string UpdateBy { get; set; }

        public DateTime CreateDate { get; set; }

        public DateTime UpdateDate { get; set; }

        public Guid? UpdateById { get; set; }

        protected bool Equals(Entity other)
        {
            if (Id == default(Guid))
                return false;
            return Id == other.Id;
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;
            return Equals((Entity) obj);
        }

        public override int GetHashCode()
        {
            return Id.GetHashCode();
        }
    }
}